packages/artifact/src/upload-artifact-http-client.ts

```diff
@@ -0,0 +1,299 @@
+        const end = offset + chunkSize - 1
+        offset += parameters.maxChunkSize
+        const chunk: NodeJS.ReadableStream = fs.createReadStream(
+          parameters.file,
+          {
+            start,
+            end,
+            autoClose: false
+          }
+        )
+
+        await uploadChunk(
+          parameters.restClient,
+          parameters.resourceUrl,
+          chunk,
+          start,
+          end,
+          fileSize
+        )
+      }
+    })
+  )
+  return fs.statSync(parameters.file).size
+}
+
+/**
+ * Uploads a chunk of an individual file to the specified resourceUrl. If the upload fails and the status code
+ * indicates a retryable status, we try to upload the chunk as well
+ * @param {HttpClient} restClient RestClient that will be making the appropriate HTTP call
+ * @param {string} resourceUrl Url of the resource that the chunk will be uploaded to
+ * @param {NodeJS.ReadableStream} data Stream of the file that will be uploaded
+ * @param {number} start Starting byte index of file that the chunk belongs to
+ * @param {number} end Ending byte index of file that the chunk belongs to
+ * @param {number} totalSize Total size of the file in bytes that is being uploaded
+ */
+async function uploadChunk(
+  restClient: HttpClient,
+  resourceUrl: string,
+  data: NodeJS.ReadableStream,
+  start: number,
+  end: number,
+  totalSize: number
+): Promise<void> {
+  console.log(
+    `Uploading chunk of size ${end -
+      start +
+      1} bytes at offset ${start} with content range: ${getContentRange(
+      start,
+      end,
+      totalSize
+    )}`
+  )
+
+  const requestOptions = getRequestOptions()
+  requestOptions['Content-Type'] = 'application/octet-stream'
+  requestOptions['Content-Length'] = totalSize
+  requestOptions['Content-Range'] = getContentRange(start, end, totalSize)
+
+  const uploadChunkRequest = async (): Promise<IHttpClientResponse> => {
+    return await restClient.sendStream('PUT', resourceUrl, data, requestOptions)
+  }
+
+  const response = await uploadChunkRequest()
+
+  if (!response.message.statusCode) {
+    console.log(response)
+    throw new Error('No Status Code returned with response')
+  }
+
+  if (isSuccessStatusCode(response.message.statusCode)) {
+    debug(
+      `Chunk for ${start}:${end} was succesfully uploaded to ${resourceUrl}`
+    )
+    return
+  }
+  if (isRetryableStatusCode(response.message.statusCode)) {
+    console.log(
+      `Received ${response.message.statusCode}, will retry chunk at offset ${start} after 10 seconds.`
+    )
+    await new Promise(resolve => setTimeout(resolve, 10000))
+    console.log(`Retrying chunk at offset ${start}`)
+
+    const retryResponse = await uploadChunkRequest()
+    if (!retryResponse.message.statusCode) {
+      console.log(retryResponse)
+      throw new Error('No Status Code returne with response')
+    }
+    if (isSuccessStatusCode(retryResponse.message.statusCode)) {
+      return
+    }
```

Comment:
What if this is not a successful status code? Shouldn't we throw an error or at least debug the response?

Reply:
Added a bunch of `console.log` statements if there is a failure. The user might not have debugging on in the event of a failure so I always want some info to show up